### PR TITLE
Control local component state with Redux

### PR DESCRIPTION
**Pull request by kumar303**
Fixes https://github.com/mozilla/addons-frontend/issues/5343

It would be great to get some feedback on this (the more eyes the better). I've only replaced `setState()` with `setUIState()` in one component, `EditableCollectionManager` to get a feel for it.

This is only a small example so it's not dramatic but here are some screencasts in the Redux tool to illustrate how switching to `setUIState()` will open up some possibilities for us.

When hiding/showing a note was controlled with `setState()`, it was completely invisible from the Redux tool (and anything else that works with Redux state). You can see that replaying all actions would not show the edit form:

<hr/>
<details>
<summary>Using setState(), the old way</summary>

addons-frontend-coll-edit-before mov

</details>
<hr/>

When notes are controlled with `setUIState()`, it is now possible to fully debug all UI state changes (including the edit form) with the Redux tool:

<hr/>
<details>
<summary>Using setUIState(), the new way</summary>

addons-frontend-coll-edit-after mov

</details>
<hr/>

The actual text input (changing the note text) is still handled by `setState()` in these examples which is why you don't see any typing.

### CODE REVIEW

File: src/core/reducers/uiState.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+      ...dispatchProps,
+      setUIState: (change: Object) => {
+        dispatch(setUIState({ id: stateProps.uiStateID, change }));
+      },
+      // This is just for testing to simulate the props you'll
+      // get after a Redux action is dispatched.
+      // Using shallowUntilTarget() and root.update() won't
+      // actually update all of the HOC wrapped components
+      // ... I guess?
+      simulateUIStateProps: ({ store }: {| store: Object |}) => {
```
Comment (willdurand) [inline src/core/reducers/uiState.js]: Any change to move this function outside of the HOC? Can't this code live in `applyUIStateChanges()` for instance?

File: src/core/reducers/uiState.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  invariant(extractId, 'extractId is required');
+  invariant(defaultState, 'defaultState is required');
+
+  const mapStateToProps = (
+    state: {| uiState: UIStateType |},
+    props: Object,
+  ) => {
+    const uiStateID =
+      props.uiStateID || generateId({ fileName, id: extractId(props) });
+    const uiState = state.uiState[uiStateID] || defaultState;
```
Comment (willdurand) [inline src/core/reducers/uiState.js]: Do we really need a `defaultState`? I wish we had a `initialState` instead, similar to:

```
constructor(props)

File: src/core/reducers/uiState.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  return (WrappedComponent) => {
+    return compose(
+      connect(
+        mapStateToProps,
+        undefined,
+        mergeProps,
+      ),
+    )(WrappedComponent);
+  };
+};
```
Comment (willdurand) [inline src/core/reducers/uiState.js]: It would be less confusing to move the HOC code in a `withUIState.js` file.

File: src/core/reducers/uiState.js (hunk 2 above)
Comment (kumar303) [reply src/core/reducers/uiState.js]: Let me try to explain the problem from the perspective of `EditableCollectionAddon`.

It was doing this to initialize state:

```js
constructor(props) {
  super(props);
  this.state = {
    editingNote: false,
  };
}
```

I introduced `withUIState({ defaultState: ... })` to mimic this but because Redux state is persistent across all component instances (because `extractId()` only looks at props), I needed this to reset the state for *every* new instance:

```js
componentWillMount() {
  this.props.setUIState({ editingNote: false });
}
```

This allowed it to function like before.

I can think of some different solutions here. Switching to `componentDidMount()` is an obvious solution but setting an initial state really isn't an async operation so let me think of some other approaches.

Comment (willdurand) [reply src/core/reducers/uiState.js]: Yes, I figured out that my code snippet above fixed only one part of the problem, but not exactly what I was highlighting. We need something to initialize the state in `uiState` before the first `render()`.

File: src/core/reducers/uiState.js (hunk 1 above)
Comment (kumar303) [reply src/core/reducers/uiState.js]: I tried this but ran into issues with the variables encapsulated by `withUIState()`. I made a late change where `uiStateID` is exposed in props so let me try again now that this computed variable is accessible.

File: src/core/reducers/uiState.js (hunk 3 above)
Comment (kumar303) [reply src/core/reducers/uiState.js]: Ah, makes sense. I'll move it.

File: src/core/reducers/uiState.js (hunk 2 above)
Comment (willdurand) [reply src/core/reducers/uiState.js]: mmh, that might work actually since the props are initialized with `defaultState`. We always have valid data injected by the HOC. By spreading the `stateProps`, the first call to `setUIState()` should add the `defaultState` overrode by `change` into redux. That should reproduce the behavior of `this.state` into the constructor as well as what `setState()` does.

File: tests/unit/amo/components/TestEditableCollectionAddon.js [hunk 4]
```diff
@@ -172,15 +186,24 @@ describe(__filename, () => {
...
     it('shows a populated notes form when the edit button icon is clicked', () => {
       const notes = 'Some notes.';
 
-      const root = render({ notes });
-      expect(root.find('.EditableCollectionAddon-notes-form')).toHaveLength(0);
-      const editButton = root.find(
-        '.EditableCollectionAddon-notes-edit-button',
-      );
-      editButton.simulate('click', createFakeEvent());
+      const root = renderAndEditNote({ notes });
```
[the remark this answers is not in the document]
Comment (kumar303) [reply tests/unit/amo/components/TestEditableCollectionAddon.js]: Oh, whoops, this test should still click the icon.

My idea was for any test that works solely in the edit-note state, they should just enter that state directly.

File: tests/unit/amo/components/TestEditableCollectionAddon.js [hunk 5]
```diff
@@ -172,15 +186,24 @@ describe(__filename, () => {
       ).toHaveLength(0);
     });
 
+    it('changes UI state when the comment icon is clicked', () => {
+      const { store } = dispatchClientMetadata();
+      const root = render({ store });
+      applyUIStateChanges({ root, store });
```
[the remark this answers is not in the document]
Comment (kumar303) [reply tests/unit/amo/components/TestEditableCollectionAddon.js]: I agree. I'm going to try and think of some other ways to solve it. I'm not totally sure why Enzyme doesn't automatically update props with Redux state but we use a similar pattern of manually applying Redux state in other tests already.

**Comment by kumar303** [comment]
@willdurand this probably isn't complete yet but it's ready for another look. Here are the main things I addressed:

- All test code resides in the helper now (no more `simulateUIStateProps`)
- All components use `componentDidMount()` to reset state which makes it behave similar to `this.setState()`

Outstanding issues:
- I'd like to not require `applyUIStateChanges()` after every state change in tests. I haven't thought of a way to solve that yet.

File: src/core/withUIState.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+import * as React from 'react';
+import invariant from 'invariant';
+import { compose } from 'redux';
+import { connect } from 'react-redux';
+
+import { normalizeFileNameId } from 'core/utils';
+import { setUIState } from 'core/reducers/uiState';
+import type { UIStateType } from 'core/reducers/uiState';
+
+type ExtractIdType = (props: Object) => string;
```
Comment (willdurand) [inline src/core/withUIState.js]: I think we used `Func` as suffix in the past: `ExtractIdFunc`.

File: src/core/withUIState.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        // props.uiState doesn't match initialState?
+        this.props.setUIState(initialState);
+      }
+
+      render() {
+        return <WrappedComponent {...this.props} />;
+      }
+    }
+
+    return compose(
```
Comment (willdurand) [inline src/core/withUIState.js]: We might not need `compose` here.

File: src/core/withUIState.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        //
+        // TODO: Optimize this by only dispatching if
+        // props.uiState doesn't match initialState?
+        this.props.setUIState(initialState);
+      }
+
+      render() {
+        return <WrappedComponent {...this.props} />;
+      }
+    }
```
Comment (willdurand) [inline src/core/withUIState.js]: It would be cool to set a display name: https://reactjs.org/docs/higher-order-components.html#convention-wrap-the-display-name-for-easy-debugging

File: tests/unit/core/testWithUIState.js [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+          <div className="overlay">
+            <button className="close-button" onClick={this.closeOverlay}>
+              Close overlay
+            </button>
+          </div>
+        );
+      }
+    }
+
+    const Overlay = compose(
```
Comment (willdurand) [inline tests/unit/core/testWithUIState.js]: You probably don't need `compose` here.

File: tests/unit/core/testWithUIState.js [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+      }),
+    )(OverlayBase);
+
+    const render = ({ id = 'some-id', ...props } = {}) => {
+      const root = shallowUntilTarget(
+        <Overlay id={id} store={store} {...props} />,
+        OverlayBase,
+      );
+      // Apply initial state change from componentWillMount.
+      applyUIStateChanges({ root, store });
```
Comment (willdurand) [inline tests/unit/core/testWithUIState.js]: Could it be because of `shallowUntilTarget`? Would a plain old `shallow` + `dive` work?

File: tests/unit/core/testWithUIState.js [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+        }),
+      )(ThingBase);
+
+      const root = shallowUntilTarget(<Thing store={store} />, ThingBase);
+
+      expect(root.instance().props.uiState).toEqual(initialState);
+    });
+
+    it('always resets initial state per component instance', () => {
+      class ThingBase extends React.Component {
```
Comment (willdurand) [inline tests/unit/core/testWithUIState.js]: I think you could probably use a functional component to reduce the amount of code in this test:

```
const ThingBase = () => <div />;
```

File: tests/unit/helpers.js [hunk 12]
```diff
@@ -611,3 +613,40 @@ export const createUserNotificationsResponse = () => {
...
+  const ownProps = root.instance().props;
+  invariant(
+    ownProps.uiStateID,
+    'uiStateID cannot be undefined; was the component wrapped in withUIState()?',
+  );
+
+  const mapStateToProps = createUIStateMapper({
+    uiStateID: ownProps.uiStateID,
+  });
+  const _stateProps = mapStateToProps(store.getState(), ownProps);
```
Comment (willdurand) [inline tests/unit/helpers.js]: Why is this variable prefixed with underscore?

File: src/core/withUIState.js (hunk 7 above)
Comment (kumar303) [reply src/core/withUIState.js]: Oh yeah, duh

File: tests/unit/core/testWithUIState.js (hunk 11 above)
Comment (kumar303) [reply tests/unit/core/testWithUIState.js]: You can't use `root.instance().props` on a stateless component because there is no instance. You also can't use `root.props()` because that only returns *rendered* props, not props passed to the stateless component's function. I think `instance()` is helpful here because we are testing the affect of *injecting* props into a component. It's not testing the rendering of a component, unlike our typical usage of Enzyme.

Comment (willdurand) [reply tests/unit/core/testWithUIState.js]: ah, I see. But the HOC should work with stateless components though, no?

File: tests/unit/core/testWithUIState.js (hunk 10 above)
Comment (kumar303) [reply tests/unit/core/testWithUIState.js]: Nope. I got the tests passing with this:

```js
const root = shallow(<EditableCollectionAddon ...>).dive().dive().dive().dive();
```

...but `applyUIStateChanges()` was still necessary.

File: tests/unit/core/testWithUIState.js (hunk 11 above)
Comment (kumar303) [reply tests/unit/core/testWithUIState.js]: Yes, the HOC will work fine with stateless components. The `root.instance().props` call is only necessary to make an assertion about the received props while testing.

File: src/core/withUIState.js (hunk 8 above)
Comment (kumar303) [reply src/core/withUIState.js]: Good idea. Done.

**Comment by kumar303** [comment]
I'm not totally happy with `applyUIStateChanges()` in tests but I think it's worth pushing forward as is for a couple reasons:

- I think in most cases it won't be necessary to call `applyUIStateChanges()`. I prefer tests that simulate a *single* user event and assert that a *single* Redux action was dispatched. This helps us think about the application in a pure way: *receive state* -> *render*. A lot of our tests rely on multiple state changes and multiple render calls.
- The patch as is will allow us to convert many `setState` calls to Redux. In a future patch we could improve the testing experience.
